presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/CachingHiveMetastore.java

```diff
@@ -938,10 +946,9 @@ public String toString()
         }
     }
 
-    private <T> KeyAndContext<T> getCachingKey(MetastoreContext context, T key)
+    private <T> KeyAndContext<T> getCachingKey(MetastoreContext metastoreContext, T key)
     {
-        MetastoreContext metastoreContext = metastoreImpersonationEnabled ? context : new MetastoreContext(NO_IMPERSONATION_USER, context.getQueryId(), context.getClientInfo(), context.getSource());
-        return new KeyAndContext<>(metastoreContext, key);
+        return new KeyAndContext<>(metastoreContext, key, metastoreImpersonationEnabled);
```

Comment:
How about we make `impersonationEnabled` part of `MetastoreContext`? This way we don't need an extra parameter here, and `KeyAndContext` can use the context information to determine its `equals` and `hashCode`.

Reply:
yeah I agree. Since `impersonation` is related to Metastore, it is better add it to `MetastoreContext`.